esp/services/ws_dali/ws_daliservice.cpp

```diff
@@ -0,0 +1,102 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef _USE_OPENLDAP
+#include "ldapsecurity.ipp"
+#endif
+
+#include "ws_daliservice.hpp"
+#include "jlib.hpp"
+#include "dautils.hpp"
+#include "dasds.hpp"
+
+#define REQPATH_EXPORTSDSDATA "/WSDali/Export"
+
+const char* daliFolder = "tempdalifiles" PATHSEPSTR;
+
+void CWSDaliEx::init(IPropertyTree* cfg, const char* process, const char* service)
+{
+    espProcess.set(process);
+}
+
+int CWSDaliSoapBindingEx::onGet(CHttpRequest* request, CHttpResponse* response)
+{
+    try
+    {
+#ifdef _USE_OPENLDAP
+        request->queryContext()->ensureSuperUser(ECLWATCH_SUPER_USER_ACCESS_DENIED, "Access denied, administrators only.");
+#endif
+        if (wsdService->isDaliDetached())
+            throw makeStringException(ECLWATCH_CANNOT_CONNECT_DALI, "Dali detached.");
+
+        StringBuffer path;
+        request->getPath(path);
+
+        if (!strnicmp(path.str(), REQPATH_EXPORTSDSDATA, sizeof(REQPATH_EXPORTSDSDATA) - 1))
+        {
+            exportSDSData(request, response);
+            return 0;
+        }
+    }
+    catch(IException* e)
+    {
+        onGetException(*request->queryContext(), request, response, *e);
+        FORWARDEXCEPTION(*request->queryContext(), e,  ECLWATCH_INTERNAL_ERROR);
+    }
+
+    return CWSDaliSoapBinding::onGet(request,response);
+}
+
+void CWSDaliSoapBindingEx::exportSDSData(CHttpRequest* request, CHttpResponse* response)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer, outFileNameWithPath;
+    VStringBuffer outFileName("sds_for_%s_%ld", request->getPeer(peer).str(), createTime);
+    outFileNameWithPath.set(daliFolder).append(outFileName);
+
+    StringBuffer path, xpath, safeReq;
+    request->getParameter("Path", path);
+    request->getParameter("Safe", safeReq);
+    Owned<IRemoteConnection> conn = connectXPathOrFile(path, strToBool(safeReq), xpath);
+    if (!conn)
+        throw makeStringException(ECLWATCH_CANNOT_CONNECT_DALI, "Failed to connect Dali.");
+
+    Owned<IFile> workingDir = createIFile(daliFolder);
+    if (!workingDir->exists())
+        workingDir->createDirectory();
+
+    Owned<IPropertyTree> root = conn->getRoot();
+    Owned<IFile> f = createIFile(outFileNameWithPath);
+    Owned<IFileIO> io = f->open(IFOcreaterw);
+    { //Force the fios to finish
+        Owned<IFileIOStream> fios = createBufferedIOStream(io);
+        toXML(root, *fios);
+    }
+
+    VStringBuffer headerStr("attachment;filename=%s", outFileName.str());
+    IEspContext* context = request->queryContext();
+    context->addCustomerHeader("Content-disposition", headerStr.str());
+
+    response->setContent(createIOStream(io));
+    response->setContentType(HTTP_TYPE_OCTET_STREAM);
+    response->send();
+
+    removeFileTraceIfFail(outFileNameWithPath);
```

Comment:
I think the file is still open at this point. Should release io 1st above.

Reply:
Are you saying that the file cannot be removed if the io is not released?

Reply:
under Windows it would fail.
under Linux, you can remove an open file, the handle will remain usable until it is closed, but you should close before deleting it.

Reply:
I see. Thanks.